### PR TITLE
[SM-650] Updating search and select all to work together properly

### PR DESCRIPTION
**Pull request by cd-bitwarden**
## Type of change

```
- [ x ] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This functionality should work the same way it works in the Password Manager vault - if a user selects All when the page is filtered, All items from the searched view should selected, not all items.

## Code changes

projects-list.component.ts, secrets-list.component.ts, service-account-list.component.ts - all updated to toggle depending on the filtered list of items, and isAllSelected is updated to look at the correct list of items.

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

**Review by Thomas-Avery — CHANGES_REQUESTED** [review]
This currently doesn't clear selections made prior to entering a value in the search field. 

This can lead to cases where a bulk delete effects previously selected values that aren't presented to the user, and select all not working as expected.

<del>We shouldn't need to make any changes to `libs/components/src/table/table-data-source.ts`. 

<del>Look at how we have a working implementation of this at `apps/web/src/app/vault/components/vault-items/vault-<del>items.component.ts` and copy the pattern used there. 

<del>Note the use of `editableItems` to keep a list of items the search string is presenting to the user.

After further review, looks like we will have to have filteredData exposed.

File: libs/components/src/table/table-data-source.ts [hunk 1]
```diff
@@ -31,6 +38,7 @@ export class TableDataSource<T> extends DataSource<T> {
   }
 
   set data(data: T[]) {
+    this.filteredData = data;
```
Comment (Hinton) [inline libs/components/src/table/table-data-source.ts:41]: We diverge slightly from the CDK, I think we can use the same implementation. https://github.com/angular/components/blob/main/src/material/table/table-data-source.ts#L92-L96

Comment (Thomas-Avery) [inline libs/components/src/table/table-data-source.ts:41]: Looking at the CDK source code this is based on, they have something like the following for `set data` and `set filter`

```
  set data(data: T[]) {
    this._data.next(data ? [...data] : []);

    if (!this._renderChangesSubscription) {
      this.filterData(data, null);
    }
  }
```

```
 set filter(filter: string) {
    this._filter.next(filter);

    if (!this._renderChangesSubscription) {
      this.filterData(this.data, filter);
    }
  }
```

With comments explaining, `filteredData` won't be populated if there isn't any re-render subscription.

@Hinton Do you see any concerns with having something like this on the setters? This is outside my area of expertise.

Comment (Hinton) [reply libs/components/src/table/table-data-source.ts:41]: No, should be fine. That's how the CDK does it. We just need to make certain to apply it to both, and to check the subscription.

File: bitwarden_license/bit-web/src/app/secrets-manager/shared/projects-list.component.ts [hunk 2]
```diff
@@ -45,15 +46,26 @@ export class ProjectsListComponent {
...
   isAllSelected() {
-    const numSelected = this.selection.selected.length;
-    const numRows = this.projects.length;
-    return numSelected === numRows;
+    if (this.selection.selected?.length > 0) {
+      const numSelected = this.selection.selected.length;
+      const numRows = this.dataSource.filter
+        ? this.dataSource.filteredData.length
+        : this.dataSource.data.length;
+      return numSelected === numRows;
```
Comment (Thomas-Avery) [inline bitwarden_license/bit-web/src/app/secrets-manager/shared/projects-list.component.ts:54]: If `filteredData` contains the filtered set of data that has been matched by the filter string, or all the data if there is no filter.

Shouldn't we be able to use just `filteredData` length here? 

```
      const numSelected = this.selection.selected.length;
      const numRows = this.dataSource.filteredData.length;
      return numSelected === numRows;
```

File: bitwarden_license/bit-web/src/app/secrets-manager/shared/projects-list.component.ts [hunk 3]
```diff
@@ -45,15 +46,26 @@ export class ProjectsListComponent {
...
-      ? this.selection.clear()
-      : this.selection.select(...this.projects.map((s) => s.id));
+    if (this.isAllSelected()) {
+      this.selection.clear();
+    } else {
+      if (this.dataSource.filter?.length > 0) {
+        this.selection.select(...this.dataSource.filteredData.map((s) => s.id));
+      } else {
+        this.selection.select(...this.dataSource.data.map((s) => s.id));
+      }
```
Comment (Thomas-Avery) [inline bitwarden_license/bit-web/src/app/secrets-manager/shared/projects-list.component.ts:67]: Same thing here, if `filteredData` has all data without the filter shouldn't we be able to just use that? 

```
  toggleAll() {
    if (this.isAllSelected()) {
      this.selection.clear();
    } else {
      this.selection.select(...this.dataSource.filteredData.map((s) => s.id));
    }
  }
```

**Review by Thomas-Avery — COMMENTED** [review]
This feedback will be applicable to all the list component changes, but just wrote it for one.

File: bitwarden_license/bit-web/src/app/secrets-manager/shared/projects-list.component.ts (hunk 2 above)
Comment (cd-bitwarden) [reply bitwarden_license/bit-web/src/app/secrets-manager/shared/projects-list.component.ts:54]: makes sense to me! 
image

**Review by Thomas-Avery — APPROVED** [review]
Looks good, thank you!

File: libs/components/src/table/table-data-source.ts [hunk 4]
```diff
@@ -31,7 +38,13 @@ export class TableDataSource<T> extends DataSource<T> {
   }
 
   set data(data: T[]) {
-    this._data.next(data ? [...data] : []);
+    data = Array.isArray(data) ? data : [];
+    this._data.next(data);
+    // Normally the `filteredData` is updated by the re-render
+    // subscription, but that won't happen if it's inactive.
+    if (!this._renderChangesSubscription) {
+      this.filterData(data, "");
```
Comment (Hinton) [inline libs/components/src/table/table-data-source.ts:46]: I expect we want to use the active filter here.

File: libs/components/src/table/table-data-source.ts [hunk 5]
```diff
@@ -48,6 +61,10 @@ export class TableDataSource<T> extends DataSource<T> {
 
   set filter(filter: string) {
     this._filter.next(filter);
+
+    if (!this._renderChangesSubscription) {
```
Comment (Hinton) [inline libs/components/src/table/table-data-source.ts:65]: Same comment as above?

File: libs/components/src/table/table-data-source.ts (hunk 4 above)
Comment (cd-bitwarden) [reply libs/components/src/table/table-data-source.ts:46]: Hey Oscar, thanks for the review! Can you explain more what you mean?

Comment (Hinton) [reply libs/components/src/table/table-data-source.ts:46]: We might have filters set that should still apply when changing the data.

File: libs/components/src/table/table-data-source.ts [hunk 6]
```diff
@@ -77,11 +95,12 @@ export class TableDataSource<T> extends DataSource<T> {
   }
 
   private filterData(data: T[], filter: string): T[] {
```
Comment (Hinton) [inline libs/components/src/table/table-data-source.ts:97]: `filter` is currently unused. We should either use it, or remove the variable.

Comment (cd-bitwarden) [reply libs/components/src/table/table-data-source.ts:97]: Thank you for noticing that! I've removed it :)

File: libs/components/src/table/table-data-source.ts [hunk 7]
```diff
@@ -65,7 +83,7 @@ export class TableDataSource<T> extends DataSource<T> {
 
   private updateChangeSubscription() {
     const filteredData = combineLatest([this._data, this._filter]).pipe(
-      map(([data, filter]) => this.filterData(data, filter))
+      map(([data, filter]) => this.filterData(data))
```
Comment (Thomas-Avery) [inline libs/components/src/table/table-data-source.ts:86]:
```suggestion
      map(([data]) => this.filterData(data))
```

We won't need filter here if we are removing it from the method's parameters.

**Review by Thomas-Avery — CHANGES_REQUESTED** [review]

**Review by Thomas-Avery — APPROVED** [review]
LGTM

**Review by Hinton — APPROVED** [review]

**Pull request merged**